src/legacy/core_plugins/table_vis/migrations.js

```diff
@@ -0,0 +1,59 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { cloneDeep, get, omit } from 'lodash';
+
+export const migrations = {
+  visualization: {
+    '7.0.0': (doc) => {
+      try {
+        const visState = JSON.parse(doc.attributes.visState);
+        if (get(visState, 'type') !== 'table') {
+          return doc; // do nothing; we only want to touch tables
```

Comment:
This seems really similar to the previous `migrateAppState` function. I was going to suggest unifying the two, but I think you're probably right to keep them separate, as the previous function can presumably morph over time, but migrations should be immutable once published.

That said, it would probably be a good idea to keep the control flow consistent between them (e.g. here, we're exiting early (me likey!) but in the previous one, there are nesting conditionals.)

Also, I wonder if the previous function `migrateAppState` is necessary at all, given that you've got this migration doing that work, you shouldn't end up with any visualizations that *need* `migrateAppState` to be applied to them, right?

Reply:
Also, right now this app state migration concept only exists here and in dashboard, but if this becomes a more prevalent problem over time, it may warrant a deeper discussion around formalizing a pattern for overriding `appState` data before it is rendered, like a hook of some kind (or dare I say "middleware"?) 😆